sql: TestStreamerTightBudget failed

sql.TestStreamerTightBudget [failed](https://tanzanite.cluster.engflow.com/invocations/default/78964d4c-58d5-4306-a532-70af90fa4fcb?testReportRun=9&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [7bb52a7d1c75d5adfdfa53e5fcff6f5e6497408f](https://github.com/cockroachdb/cockroach/commits/7bb52a7d1c75d5adfdfa53e5fcff6f5e6497408f):


```
        	            	----
        	            	planning time: 302µs
        	            	execution time: 7ms
        	            	distribution: full
        	            	vectorized: true
        	            	rows decoded from KV: 10 (5.0 MiB, 10 gRPC calls)
        	            	cumulative time spent in KV: 6ms
        	            	maximum memory usage: 5.3 MiB
        	            	network usage: 0 B (0 messages)
        	            	sql cpu time: 5ms
        	            	isolation level: serializable
        	            	priority: normal
        	            	quality of service: regular
        	            	
        	            	• group (scalar)
        	            	│ nodes: n1
        	            	│ actual row count: 1
        	            	│ sql cpu time: 23µs
        	            	│
        	            	└── • render
        	            	    │
        	            	    └── • index join (streamer)
        	            	        │ nodes: n1
        	            	        │ actual row count: 5
        	            	        │ KV time: 6ms
        	            	        │ KV contention time: 0µs
        	            	        │ KV rows decoded: 5
        	            	        │ KV bytes read: 5.0 MiB
        	            	        │ KV gRPC calls: 9
        	            	        │ estimated max memory allocated: 4.6 MiB
        	            	        │ estimated max sql temp disk usage: 0 B
        	            	        │ sql cpu time: 5ms
        	            	        │ table: t@t_pkey
        	            	        │
        	            	        └── • scan
        	            	              nodes: n1
        	            	              actual row count: 5
        	            	              KV time: 115µs
        	            	              KV contention time: 0µs
        	            	              KV rows decoded: 5
        	            	              KV bytes read: 168 B
        	            	              KV gRPC calls: 1
        	            	              estimated max memory allocated: 10 KiB
        	            	              sql cpu time: 6µs
        	            	              missing stats
        	            	              table: t@t_k_idx
        	            	              spans: [/1 - /1]
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestStreamerTightBudget1878394043
--- FAIL: TestStreamerTightBudget (0.54s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=9</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamerTightBudget.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-36274

--- a/pkg/sql/explain_bundle_test.go
+++ b/pkg/sql/explain_bundle_test.go
@@ -466,7 +466,7 @@ func getBundleDownloadURL(t *testing.T, text string) string {
 	return url
 }
 
-func downloadAndUnzipBundle(t *testing.T, url string) *zip.Reader {
+func downloadBundle(t *testing.T, url string, dest io.Writer) {
 	httpClient := httputil.NewClientWithTimeout(30 * time.Second)
 	// Download the zip to a BytesBuffer.
 	resp, err := httpClient.Get(context.Background(), url)
@@ -474,8 +474,13 @@ func downloadAndUnzipBundle(t *testing.T, url string) *zip.Reader {
 		t.Fatal(err)
 	}
 	defer resp.Body.Close()
+	_, _ = io.Copy(dest, resp.Body)
+}
+
+func downloadAndUnzipBundle(t *testing.T, url string) *zip.Reader {
+	// Download the zip to a BytesBuffer.
 	var buf bytes.Buffer
-	_, _ = io.Copy(&buf, resp.Body)
+	downloadBundle(t, url, &buf)
 
 	unzip, err := zip.NewReader(bytes.NewReader(buf.Bytes()), int64(buf.Len()))
 	if err != nil {
@@ -483,6 +488,19 @@ func downloadAndUnzipBundle(t *testing.T, url string) *zip.Reader {
 		t.Fatal(err)
 	}
 	return unzip
+}
+
+func readUnzippedFile(t *testing.T, f *zip.File) string {
+	r, err := f.Open()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer r.Close()
+	bytes, err := io.ReadAll(r)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return string(bytes)
 }
 
 // checkBundle searches text strings for a bundle URL and then verifies that the
@@ -515,17 +533,7 @@ func checkBundle(
 		}
 		files = append(files, f.Name)
 
-		r, err := f.Open()
-		if err != nil {
-			t.Fatal(err)
-		}
-		defer r.Close()
-		bytes, err := io.ReadAll(r)
-		if err != nil {
-			t.Fatal(err)
-		}
-		contents := string(bytes)
-
+		contents := readUnzippedFile(t, f)
 		if !expectErrors && strings.Contains(contents, "-- error") {
 			t.Errorf(
 				"expected no errors in %s, file contents:\n%s",
@@ -658,17 +666,9 @@ func TestExplainClientTime(t *testing.T) {
 	var contents string
 	for _, f := range unzip.File {
 		if f.Name == "plan.txt" {
-			r, err := f.Open()
-			if err != nil {
-				t.Fatal(err)
-			}
-			defer r.Close()
-			bytes, err := io.ReadAll(r)
-			if err != nil {
-				t.Fatal(err)
-			}
-			contents = string(bytes)
+			contents = readUnzippedFile(t, f)
 			t.Logf("contents of plan.txt\n%s", contents)
+			break
 		}
 	}
 

--- a/pkg/sql/mem_limit_test.go
+++ b/pkg/sql/mem_limit_test.go
@@ -13,6 +13,8 @@ package sql
 import (
 	"context"
 	"fmt"
+	"os"
+	"path/filepath"
 	"regexp"
 	"strconv"
 	"strings"
@@ -22,6 +24,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/datapathutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -161,6 +164,7 @@ func TestStreamerTightBudget(t *testing.T) {
 	// Streamer isn't hitting the root budget exceeded error.
 	s, db, _ := serverutils.StartServer(t, base.TestServerArgs{
 		SQLMemoryPoolSize: 1 << 30, /* 1GiB */
+		Insecure:          true,
 	})
 	defer s.Stopper().Stop(context.Background())
 	sqlDB := sqlutils.MakeSQLRunner(db)
@@ -185,9 +189,9 @@ func TestStreamerTightBudget(t *testing.T) {
 	sqlDB.Exec(t, fmt.Sprintf("SET distsql_workmem = '%dB'", blobSize))
 
 	// Perform an index join to read the blobs.
-	query := "EXPLAIN ANALYZE SELECT sum(length(blob)) FROM t@t_k_idx WHERE k = 1"
+	query := "SELECT sum(length(blob)) FROM t@t_k_idx WHERE k = 1"
 	maximumMemoryUsageRegex := regexp.MustCompile(`maximum memory usage: (\d+\.\d+) MiB`)
-	rows := sqlDB.QueryStr(t, query)
+	rows := sqlDB.QueryStr(t, "EXPLAIN ANALYZE (VERBOSE) "+query)
 	// Build pretty output for an error message in case we need it.
 	var sb strings.Builder
 	for _, row := range rows {
@@ -202,9 +206,45 @@ func TestStreamerTightBudget(t *testing.T) {
 			// by the ColIndexJoin when the blob is copied into the columnar
 			// batch). We allow for 0.1MiB for other memory usage in the query.
 			maxAllowed := 2.1
-			require.GreaterOrEqualf(t, maxAllowed, usage, "unexpectedly high memory usage\n----\n%s", sb.String())
-			return
+			if maxAllowed >= usage {
+				return
+			}
+			// Get a stmt bundle for this query that might help in debugging the
+			// failure.
+			rows = sqlDB.QueryStr(t, "EXPLAIN ANALYZE (DEBUG) "+query)
+			url := getBundleDownloadURL(t, fmt.Sprint(rows))
+			// First check whether the bundle experienced the elevated memory
+			// usage too.
+			unzip := downloadAndUnzipBundle(t, url)
+			for _, f := range unzip.File {
+				switch f.Name {
+				case "plan.txt":
+					contents := readUnzippedFile(t, f)
+					sb.WriteString("\n\n\nbundle plan.txt:\n\n\n")
+					sb.WriteString(contents)
+					// Check that the memory usage is still elevated.
+					if matches := maximumMemoryUsageRegex.FindStringSubmatch(contents); len(matches) == 0 {
+						t.Fatalf("unexpectedly didn't find a match for maximum memory usage\n%s", sb.String())
+					} else {
+						bundleUsage, err := strconv.ParseFloat(matches[1], 64)
+						require.NoError(t, err)
+						if maxAllowed >= bundleUsage {
+							// Memory usage is as expected in the bundle - do
+							// not fail the test since we don't have any
+							// information to help in understanding the original
+							// high memory usage.
+							return
+						}
+					}
+				}
+			}
+			// NB: we're not using t.TempDir() because we want these to survive
+			// on failure.
+			f, err := os.Create(filepath.Join(datapathutils.DebuggableTempDir(), "bundle.zip"))
+			require.NoError(t, err)
+			downloadBundle(t, url, f)
+			t.Fatalf("unexpectedly high memory usage\n----\n%s", sb.String())
 		}
 	}
-	t.Fatal("unexpectedly didn't find a match for maximum memory usage")
+	t.Fatalf("unexpectedly didn't find a match for maximum memory usage\n%s", sb.String())
 }

